builder/config.go

```diff
@@ -114,13 +115,49 @@ func getWarningsForObsoleteFields(reader io.Reader) ([]string, error) {
 }
 
 // parseConfig reads a builder configuration from reader and resolves relative buildpack paths using `relativeToDir`
-func parseConfig(reader io.Reader, relativeToDir string) (Config, error) {
+func parseConfig(reader io.Reader, relativeToDir, path string) (Config, error) {
 	var builderConfig Config
-	if _, err := toml.DecodeReader(reader, &builderConfig); err != nil {
+	tomlMetadata, err := toml.DecodeReader(reader, &builderConfig)
+	if err != nil {
 		return Config{}, errors.Wrap(err, "decoding toml contents")
 	}
 
+	undecodedKeys := tomlMetadata.Undecoded()
+	if len(undecodedKeys) > 0 {
+		unusedKeys := map[string]interface{}{}
+		for _, key := range undecodedKeys {
+			keyName := key.String()
+
+			parent := strings.Split(keyName, ".")[0]
+
+			if _, ok := unusedKeys[parent]; !ok {
+				unusedKeys[keyName] = nil
+			}
+		}
+
+		var errorKeys []string
+		for errorKey := range unusedKeys {
+			errorKeys = append(errorKeys, style.Symbol(errorKey))
+		}
+
+		pluralizedElement := "element"
+		if len(errorKeys) > 1 {
+			pluralizedElement += "s"
+		}
+
+		return Config{}, errors.Errorf("unknown configuration %s %s in %s",
+			pluralizedElement,
+			strings.Join(errorKeys, ", "),
+			style.Symbol(path),
+		)
+	}
```

Comment:
This code is duplicated in buildpackage/config_reader.go - could it be extracted into some sort of helper method?

Reply:
Yeah, working on it.